pkg/controller/job/job_controller.go

```diff
@@ -1682,6 +1698,69 @@ func recordJobPodFinished(job *batch.Job, oldCounters batch.JobStatus) {
 	metrics.JobPodsFinished.WithLabelValues(completionMode, metrics.Failed).Add(float64(diff))
 }
 
+func getFailedPods(pods []*v1.Pod) []*v1.Pod {
+	var result []*v1.Pod
+	for _, p := range pods {
+		if p.Status.Phase == v1.PodFailed {
+			result = append(result, p)
+		}
+	}
+	return result
+}
+
+func findMatchingBackoffPolicyRule(job *batch.Job, failedPod *v1.Pod) *v1.BackoffPolicyRule {
+	for _, backoffPolicyRule := range job.Spec.BackoffPolicy.Rules {
+		if backoffPolicyRule.OnExitCodes != nil {
+			if isOnExitCodesMatching(failedPod, backoffPolicyRule.OnExitCodes) {
+				return &backoffPolicyRule
+			}
+		}
+	}
+	return nil
+}
+
+func isOnExitCodesMatching(failedPod *v1.Pod, requirement *v1.BackoffPolicyOnExitCodesRequirement) bool {
+	containerStatuses := getAllContainers(failedPod.Status)
+	klog.InfoS("_______ containerStatuses", "containerStatuses", containerStatuses)
+	for _, containerStatus := range containerStatuses {
+		if requirement.ContainerName == nil || *requirement.ContainerName == containerStatus.Name {
+			if isOnExitCodesOperatorMatching(containerStatus.State.Terminated.ExitCode, requirement.Operator, requirement.Values) {
+				return true
+			}
+		}
+		if requirement.ContainerName != nil && *requirement.ContainerName == containerStatus.Name {
+			return false
+		}
+	}
+	return false
+}
+
+func isOnExitCodesOperatorMatching(exitCode int32, operator v1.BackoffPolicyOnExitCodesOperator, values []int32) bool {
+	if operator == v1.BackoffPolicyOnExitCodesOpIn {
```

Comment:
in the codebase, when doing operators, it's more common to use switch.

Reply:
updated